### PR TITLE
[Artifacts] Parse producer URI when retaining producer

### PR DESCRIPTION
**Pull request by TomerShor**
When an artifact is logged in a run, its producer doesn't necessarily contain a `project` and `tag` properties, but it does contain a `uri` (as in the context object).

When importing artifacts that were created in a run, we retain the run as the artifact producer.
However, the URI was ignored and we only looked for a `project` and a `tag`, causing the retained producer to not have a UID.

In this PR we first look for the project and tag on the producer itself, and if not found we try to parse its URI.

Resolves https://iguazio.atlassian.net/browse/ML-4134

### CODE REVIEW

File: mlrun/artifacts/manager.py [hunk 1]
```diff
@@ -76,6 +77,23 @@ def get_meta(self) -> dict:
     def uid(self):
         return None
 
+    @staticmethod
+    def parse_uri(uri):
+        """parse artifact producer's uri to project, uid, iteration"""
+        uri_pattern = r"^((?P<project>.*)/)?(?P<uid>.*?)(\-(?P<iteration>.*?))?$"
```
Comment (liranbg) [inline mlrun/artifacts/manager.py]: move to the piled-on `regex.py`

File: mlrun/artifacts/manager.py [hunk 2]
```diff
@@ -76,6 +77,23 @@ def get_meta(self) -> dict:
     def uid(self):
         return None
 
+    @staticmethod
+    def parse_uri(uri):
```
Comment (liranbg) [inline mlrun/artifacts/manager.py]:
```suggestion
    def parse_uri(uri) -> typing.Tuple[str, str, str]:
```

**Review by liranbg — COMMENTED** [review]
minor syntax suggestions

File: mlrun/artifacts/manager.py [hunk 3]
```diff
@@ -76,6 +77,23 @@ def get_meta(self) -> dict:
...
+        """parse artifact producer's uri to project, uid, iteration"""
+        uri_pattern = r"^((?P<project>.*)/)?(?P<uid>.*?)(\-(?P<iteration>.*?))?$"
+        match = re.match(uri_pattern, uri)
+        if not match:
+            raise ValueError(
+                "Uri not in supported format [<project>/]<key>[#<iteration>][:<tag>][@<tree>]"
```
Comment (alonmr) [inline mlrun/artifacts/manager.py]: Update the error message

**Review by alonmr — COMMENTED** [review]
~please also change in `server.api.crud.runs.Runs.get_run` to use the new utility~

**Review by alonmr — APPROVED** [review]

**Pull request merged**